Fix copysing in ONB, fix ray_intersection_list(box)

stevelonny opened the pull request:
Fix for issues #24 and #26 

Open a pull request so that i can iterate and verify thoroughly each if/else check and edge cases, and update tests as needed.


- [x] Check surface points from the boxes
- [x] Check various 1.0/dir and verify their expected behaviour
- [x] Update test:
  - [x] check various ray_intersection_list and their results
- [x] Update changelog

giogala commented:
I report another bug in the `ray_intersection_list` methods. Here I report the one I found for `CSGIntersection` but I think it can be the same somewhere else:
```julia
function ray_intersection(I::CSGIntersection, ray::Ray)
    inv_ray = inverse(I.Tr)(ray) # ray in the un-transformed intersection system
    HR_list = ray_intersection_list(I, inv_ray)
    return isnothing(HR_list) ? nothing : HR_list[1]
end

[...]

function ray_intersection_list(I::CSGIntersection, ray::Ray)
    inv_ray = inverse(I.Tr)(ray) # ray in the un-transformed intersection system
    # collect the hrs of both shapes
    HR1_list = ray_intersection_list(I.Sh1, inv_ray)
    if isnothing(HR1_list)
        return nothing
    end
```
The method `ray_intersection` transform the ray and then calls `ray_intersection_list` on the transformed ray, however the latter methods, again, before doing anything else transform the ray got in input that was the transformed one in `ray_intersection`

stevelonny commented:
I added test to both Sphere and Box that mimics the already existing one.
I changed the behavior of ray_intersection_list for the sphere such that if the the ray originate from inside, it returns nothing. This way I stopped myself from constructing useless if/else branches to determine which of the two intersection points has to be save, while it would be useless to have just a single HitRecord for a CSG figure to work with.

stevelonny commented:
There are no checks if `inv_ray.dir` is a (0,0,0) vector, however we can assume with good confidence that the ray direction will never be zero in any circumstance in our code.
